### PR TITLE
Content cards base library setup

### PR DESCRIPTION
**Pull request by navratan-soni**
## Description

Added base classes for content-cards feature in Android.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: code/aepuitemplates/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    enableSpotless = true
+    enableCheckStyle = true
+    enableDokkaDoc = true
+
+    publishing {
+        mavenRepoName = aepUiTemplatesMavenRepoName
+        mavenRepoDescription = aepUiTemplatesMavenRepoDescription
+        gitRepoName = "aepsdk-messaging-android"
+        addCoreDependency(mavenCoreVersion)
+    }
```
Comment (spoorthipujariadobe) [inline code/aepuitemplates/build.gradle.kts:25]: Same comment here. Since we're not publishing this artifact, can we leave this part out?

File: code/aepuitemplates/src/main/java/com/adobe/marketing/mobile/aepuitemplates/SmallImageTemplate.kt [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+    /** The description for the small image template. */
+    val description: String = ""
+
+    /**
+     * Returns the type of this template, which is [AepUiTemplateType.SMALL_IMAGE].
+     *
+     * @return A string representing the type of the template.
+     */
+    override fun getType() = AepUiTemplateType.SMALL_IMAGE.typeName
```
Comment (spoorthipujariadobe) [inline code/aepuitemplates/src/main/java/com/adobe/marketing/mobile/aepuitemplates/SmallImageTemplate.kt:25]: Wondering if we should change `getType` to return `AepUiTemplateType` instead of string. Thoughts?

File: code/gradle.properties [hunk 3]
```diff
@@ -33,4 +33,14 @@ mavenEdgeIdentityVersion=3.0.0
...
+
+aepComposeUiModuleName = aepcomposeui
+aepComposeUiVersion=1.0.0
+aepComposeUiMavenRepoName = AdobeMobileAEPComposeUiSdk
+aepComposeUiMavenRepoDescription = Android AEP Compose UI library for Adobe Mobile Marketing
+
+aepUiTemplatesModuleName = aepuitemplates
+aepUiTemplatesVersion=1.0.0
+aepUiTemplatesMavenRepoName = AdobeMobileAEPUiTemplates
+aepUiTemplatesMavenRepoDescription = Android AEP UI templates library for Adobe Mobile Marketing
```
Comment (spoorthipujariadobe) [inline code/gradle.properties:46]: Inclined to remove these since we're not publishing the artifact

File: code/aepuitemplates/src/main/java/com/adobe/marketing/mobile/aepuitemplates/utils/AepUiTemplateType.kt [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+ * Enum class representing different types of UI templates in AEP.
+ *
+ * @property typeName The name of the template type.
+ */
+enum class AepUiTemplateType(val typeName: String) {
+    /** Represents a small image template type. */
+    SMALL_IMAGE("SmallImage"),
+
+    /** Represents a large image template type. */
+    LARGE_IMAGE("LargeImage")
```
Comment (spoorthipujariadobe) [inline code/aepuitemplates/src/main/java/com/adobe/marketing/mobile/aepuitemplates/utils/AepUiTemplateType.kt:13]: Can we add the enum when we add support for Large Image card in the future?

File: code/aepcomposeui/build.gradle.kts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        mavenRepoName = aepComposeUiMavenRepoName
+        mavenRepoDescription = aepComposeUiMavenRepoDescription
+        gitRepoName = "aepsdk-messsaging-android"
+        addCoreDependency(mavenCoreVersion)
+    }
+}
+
+dependencies {
+    api(project(":aepuitemplates"))
+    implementation("com.adobe.marketing.mobile:core:$mavenCoreVersion")
```
Comment (spoorthipujariadobe) [inline code/aepcomposeui/build.gradle.kts:30]: Why do we need Core SDK dependency here? Are there any utilities we're using from Core?

File: code/aepcomposeui/src/main/java/com/adobe/marketing/mobile/aepcomposeui/interactions/UIEevent.kt [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    data class Dismiss<T : AEPUITemplate, S : AEPUIState>(val _aepui: AEPUI<T, S>) : UIEvent<T, S>
+
+    /**
+     * Event that represents the expansion of a UI element.
+     *
+     * @param T The type of UI template associated with the expand event.
+     * @param S The type of UI state associated with the expand event.
+     * @property _aepui The AEPUI associated with the expand event.
+     */
+    data class Expand<T : AEPUITemplate, S : AEPUIState>(val _aepui: AEPUI<T, S>) : UIEvent<T, S>
```
Comment (prudrabhat) [inline code/aepcomposeui/src/main/java/com/adobe/marketing/mobile/aepcomposeui/interactions/UIEevent.kt:48]: This needs to be changed to `Display`, `Dismiss` and `Interact` (with an `Action` to specify  the type of interaction)

File: code/aepcomposeui/build.gradle.kts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    enableSpotless = true
+    enableCheckStyle = true
+    enableDokkaDoc = true
+
+    publishing {
+        mavenRepoName = aepComposeUiMavenRepoName
+        mavenRepoDescription = aepComposeUiMavenRepoDescription
+        gitRepoName = "aepsdk-messsaging-android"
+        addCoreDependency(mavenCoreVersion)
+    }
```
[the remark this answers is not in the document]
Comment (prudrabhat) [reply code/aepcomposeui/build.gradle.kts:25]: Do we know for sure that we can leverage `api(...)` as a way to allow transitive dependency delivery when distributing it via Maven without releasing the internal artifact to Maven? I don't see a reason to create modules yet.  Can we make the templates and compose-ui modules a part of the Messaging extension itself with the desired package names?

File: code/aepcomposeui/src/main/java/com/adobe/marketing/mobile/aepcomposeui/contentprovider/AEPUIContentProvider.kt [hunk 8]
```diff
@@ -0,0 +1,21 @@
+package com.adobe.marketing.mobile.aepcomposeui.contentprovider
+
+import com.adobe.marketing.mobile.aepuitemplates.AEPUITemplate
+import kotlinx.coroutines.flow.Flow
+
+/**
+ * Interface to provide content for the UI and a backing tracker for analytics.
+ * This can be treated as a strategy to provide content and tracking for the UI.
```
Comment (spoorthipujariadobe) [inline code/aepcomposeui/src/main/java/com/adobe/marketing/mobile/aepcomposeui/contentprovider/AEPUIContentProvider.kt:8]: `and a backing tracker for analytics.` & `and tracking for the UI.` are outdated comments. AEPUIContentProvider no longer provides a tracker. Please modify this description to reflect the current functionality

File: code/aepcomposeui/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/AEPUI.kt [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+
+import androidx.compose.runtime.mutableStateOf
+import com.adobe.marketing.mobile.aepuitemplates.AEPUITemplate
+import com.adobe.marketing.mobile.aepuitemplates.SmallImageTemplate
+import com.adobe.marketing.mobile.aepcomposeui.aepui.state.AEPUIState
+import com.adobe.marketing.mobile.aepcomposeui.aepui.state.SmallImageUIState
+
+
+/**
+ * Represents a UI component that can be rendered in the AepUI Engine. Binds a template data with a tracking observer.
```
Comment (spoorthipujariadobe) [inline code/aepcomposeui/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/AEPUI.kt:11]: There is no AEPUIEngine anymore. Neither is there a tracking observer. Please update the description here

**Comment by prudrabhat** [comment]
Can you please create a `feature/content-cards` branch base this PR agains it.

File: code/aepcomposeui/build.gradle.kts (hunk 5 above)
Comment (prudrabhat) [reply code/aepcomposeui/build.gradle.kts:30]: I believe we need this for NetworkService. This will become implicit when we move this to a Messaging sub package.

File: code/aepcomposeui/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/state/AEPUIState.kt [hunk 10]
```diff
@@ -0,0 +1,6 @@
+package com.adobe.marketing.mobile.aepcomposeui.aepui.state
+
+/**
+ * Sealed interface representing different states of an AEP UI.
```
Comment (spoorthipujariadobe) [inline code/aepcomposeui/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/state/AEPUIState.kt:4]: What does `different states of an AEP UI` mean? An implementation of AEP UI can have only a single state. Can you change to something to this effect?
`Represents the properties backing AEP UI that can be changed on the interaction with the UI.`

File: code/aepcomposeui/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/state/SmallImageUIState.kt [hunk 11]
```diff
@@ -0,0 +1,10 @@
+package com.adobe.marketing.mobile.aepcomposeui.aepui.state
+
+/**
+ * Data class representing the state of a small image UI.
+ *
+ * @property title The title of the small image UI.
+ * @property dismissed Indicates whether the UI has been dismissed.
+ */
+data class SmallImageUIState(val title: String?, val dismissed: Boolean = false) : AEPUIState
```
Comment (prudrabhat) [inline code/aepcomposeui/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/state/SmallImageUIState.kt:9]: Wondering if we want to explore a top level _AepCardState_ interface that holds different behaviors like dismissed, selected etc.
```
  interface AepCardState {
    val dismissed: Boolean,
    val selected:  Boolean
}

data class SmallImageUIState(override var dismissed: Boolean = ...) : AepCardState

```

**Review by prudrabhat — COMMENTED** [review]
A few high level comments that apply to the PR:
-  Can you please create a feature/content-cards branch base this PR against it
- I don't see a reason to create modules yet. Can we make the templates and compose-ui modules a part of the Messaging extension itself with the desired package names?
- Let us align with Kotlin convention for [acronyms](https://kotlinlang.org/docs/coding-conventions.html#choose-good-names) - `AepUI`, `UI`
- Please ensure that the Kdocs for components you have in the PR reflect accurately.

File: code/aepcomposeui/build.gradle.kts (hunk 7 above)
[the remark this answers is not in the document]
Comment (spoorthipujariadobe) [reply code/aepcomposeui/build.gradle.kts:25]: Agreed. I don't think these need to be separate modules

File: code/aepcomposeui/src/main/java/com/adobe/marketing/mobile/aepcomposeui/interactions/UIEevent.kt [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+
+import com.adobe.marketing.mobile.aepuitemplates.AEPUITemplate
+import com.adobe.marketing.mobile.aepcomposeui.aepui.AEPUI
+import com.adobe.marketing.mobile.aepcomposeui.aepui.state.AEPUIState
+
+/**
+ * Sealed interface representing different types of UI events in AEP.
+ *
+ * @param T The type of UI template associated with the event.
+ * @param S The type of UI state associated with the event.
```
Comment (spoorthipujariadobe) [inline code/aepcomposeui/src/main/java/com/adobe/marketing/mobile/aepcomposeui/interactions/UIEevent.kt:11]: `type of UI template/UI state` does not really describe how the template or state is related to the event produced here. Can we change it to something like this `AepUITemplate instance/ AepUiState instance backing the Composable on which the event has occurred`

File: code/aepcomposeui/src/main/java/com/adobe/marketing/mobile/aepcomposeui/observers/AEPUIEventObserver.kt [hunk 13]
```diff
@@ -0,0 +1,15 @@
+package com.adobe.marketing.mobile.aepcomposeui.observers
+
+import com.adobe.marketing.mobile.aepcomposeui.interactions.UIEvent
+
+/**
+ * Interface for observing AEP UI events.
+ */
+interface AEPUIEventObserver {
+    /**
+     * Handles an AEP UI event.
```
Comment (spoorthipujariadobe) [inline code/aepcomposeui/src/main/java/com/adobe/marketing/mobile/aepcomposeui/observers/AEPUIEventObserver.kt:10]: Can we mention when this method is called? Something like
`Called on changes to UI templates`

And can we make it explicit on who handles the event? We can say `Implementers can provide logic to handle event based on event type`

File: code/aepcomposeui/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/AEPUI.kt [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+ * This is a sealed interface that can be implemented by different UI components like [SmallImageAEPUI], [LargeImageAEPUI], etc.
+ * This allows restricting the type of template and observer that can be associated with a UI component, which can be later
+ * used in the app to render the UI component via UIComposeExtensions.
```
Comment (spoorthipujariadobe) [inline code/aepcomposeui/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/AEPUI.kt:14]: This was added in the POC for our internal understanding only. It should not be a part of the Kdocs

Comment (spoorthipujariadobe) [inline code/aepcomposeui/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/AEPUI.kt:14]: This was added in the POC for our internal understanding only. It should not be a part of the Kdocs

File: code/aepcomposeui/build.gradle.kts (hunk 5 above)
Comment (navratan-soni) [reply code/aepcomposeui/build.gradle.kts:30]: I added this to get dependency on compose in submodule

File: code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/AepUI.kt [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+package com.adobe.marketing.mobile.aepcomposeui.aepui
+
+import com.adobe.marketing.mobile.aepcomposeui.aepui.state.AepUIState
+import com.adobe.marketing.mobile.aepuitemplates.AepUITemplate
+
+/**
+ * Represents a UI component that can be rendered using AEP UI templates.
+ * This is a sealed interface that can be implemented by different UI components like [SmallImageCardAepUi], [LargeImageCardAepUi], etc.
+ * It allows restricting the type of template and state that can be associated with a UI component, ensuring consistency
+ * when rendering and managing UI components across the app.
```
Comment (spoorthipujariadobe) [inline code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/AepUI.kt:21]: Can we please remove the reasoning of why this is a sealed interface? We added this for our internal understanding only, no reason to have it in the public Kdocs

File: code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/AepUI.kt [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+     * Retrieves the current state of the UI component.
+     * The state defines the dynamic behaviors or status of the UI component.
+     *
+     * @return The current state of type [S] associated with the UI component.
+     */
+    fun getState(): S
+
+    /**
+     * Updates the state of the UI component with a new state.
+     * The updated state will reflect in the UI when it's rendered.
```
Comment (spoorthipujariadobe) [inline code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/AepUI.kt:46]: Can we reword it to something like this - `Updates to the state will cause the UI to be recomposed`

File: code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/SmallImageAepUI.kt [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+    override fun updateState(newState: SmallImageUIState) {
+        _state.value = newState
+    }
+
+    /**
+     * Retrieves the template associated with the small image UI.
+     *
+     * @return The small image template.
+     */
+    override fun getTemplate(): com.adobe.marketing.mobile.aepuitemplates.SmallImageTemplate {
```
Comment (spoorthipujariadobe) [inline code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/SmallImageAepUI.kt:44]: nit: Let's add an import statement for`SmallImageTemplate`

File: code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/state/SmallImageUIState.kt [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+  OF ANY KIND, either express or implied. See the License for the specific language
+  governing permissions and limitations under the License.
+*/
+
+package com.adobe.marketing.mobile.aepcomposeui.aepui.state
+
+/**
+ * Data class representing the state of a small image UI.
+ *
+ * @property title The title of the small image UI.
```
Comment (spoorthipujariadobe) [inline code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/state/SmallImageUIState.kt:17]: `title` is not a changeable state. Let's remove it in a future PR when the call is going to be implemented fully

File: code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/contentprovider/AepUIContentProvider.kt [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+
+package com.adobe.marketing.mobile.aepcomposeui.contentprovider
+
+import com.adobe.marketing.mobile.aepuitemplates.AepUITemplate
+import kotlinx.coroutines.flow.Flow
+
+/**
+ * Represents a strategy to provide content for rendering the UI.
+ * Allows components like extensions which provide content to expose an implementation
+ * for the app to use.
```
Comment (spoorthipujariadobe) [inline code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/contentprovider/AepUIContentProvider.kt:20]: Not sure if we should describe details of how extensions will consume it in public docs. Can we keep it simple `Responsible for retrieving and refreshing data as required by the UI`

File: code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/interactions/UIEvent.kt [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+*/
+
+package com.adobe.marketing.mobile.aepcomposeui.interactions
+
+import com.adobe.marketing.mobile.aepcomposeui.aepui.AepUI
+import com.adobe.marketing.mobile.aepcomposeui.aepui.state.AepUIState
+import com.adobe.marketing.mobile.aepuitemplates.AepUITemplate
+
+/**
+ * Sealed interface representing different types of UI events in AEP.
```
Comment (spoorthipujariadobe) [inline code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/interactions/UIEvent.kt:19]: nit: `..representing different types of UI events that can be produced by supported UI templates`

File: code/messaging/src/main/java/com/adobe/marketing/mobile/aepuitemplates/SmallImageTemplate.kt [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+*/
+
+package com.adobe.marketing.mobile.aepuitemplates
+
+import com.adobe.marketing.mobile.aepuitemplates.utils.AepUITemplateType
+
+/**
+ * Class representing a small image template, which implements the [AepUITemplate] interface.
+ *
+ * This class contains properties for an image URL, title, and description.
```
Comment (spoorthipujariadobe) [inline code/messaging/src/main/java/com/adobe/marketing/mobile/aepuitemplates/SmallImageTemplate.kt:19]: This needs to be changed in future PRs to reflect the actual data fields needed to represent small image card

File: code/aepcomposeui/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/components/SmallImageCard.kt [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+
+    LaunchedEffect(key1 = Unit) {
+        observer?.onEvent(UIEvent.Display(ui))
+    }
+
+    DisposableEffect(key1 = Unit) {
+        onDispose {
+            observer?.onEvent(UIEvent.Dismiss(ui))
+        }
+    }
```
[the remark this answers is not in the document]
Comment (navratan-soni) [reply code/aepcomposeui/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/components/SmallImageCard.kt]: Not sure about this.
Right now this class was added just for placeholder. We can revisit this when actual implementation is done.

File: code/messaging/build.gradle.kts [hunk 23]
```diff
@@ -21,6 +21,7 @@ aepLibrary {
     namespace = "com.adobe.marketing.mobile.messaging"
     enableSpotless = true
     enableCheckStyle = true
+    enableDokkaDoc = true
```
Comment (spoorthipujariadobe) [inline code/messaging/build.gradle.kts:24]: The make file also needs to be changed to use dokka. 
https://github.com/adobe/aepsdk-messaging-android/blob/main/Makefile#L40 -> https://github.com/adobe/aepsdk-core-android/blob/main/Makefile#L67

File: code/messaging/src/main/java/com/adobe/marketing/mobile/aepuitemplates/SmallImageTemplate.kt (hunk 21 above)
Comment (navratan-soni) [reply code/messaging/src/main/java/com/adobe/marketing/mobile/aepuitemplates/SmallImageTemplate.kt:19]: Sure when proper implementation of the class is added, will change the kdocs which reflects the same

**Review by spoorthipujariadobe — APPROVED** [review]
